File: modules/on-device-translation/android/src/main/java/expo/modules/ondevicetranslation/OnDeviceTranslationModule.kt
package expo.modules.ondevicetranslation

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import com.google.mlkit.nl.translate.TranslateLanguage
import com.google.mlkit.nl.translate.Translation
import com.google.mlkit.nl.translate.TranslatorOptions
import com.google.mlkit.common.model.DownloadConditions

class OnDeviceTranslationModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("OnDeviceTranslation")
    AsyncFunction("translate") { text: String, source: String, target: String, promise: Promise ->
      val from = TranslateLanguage.fromLanguageTag(source)
      val to = TranslateLanguage.fromLanguageTag(target)
      if (from == null || to == null) {
        promise.reject("UNSUPPORTED_LANGUAGE", "Deze taal wordt niet ondersteund op dit toestel", null)
      } else {
        val translator = Translation.getClient(TranslatorOptions.Builder().setSourceLanguage(from).setTargetLanguage(to).build())
        translator.downloadModelIfNeeded(DownloadConditions.Builder().requireWifi().build())
          .addOnSuccessListener {
            translator.translate(text)
              .addOnSuccessListener { value -> promise.resolve(value); translator.close() }
              .addOnFailureListener { error -> promise.reject("TRANSLATION_FAILED", error.message, error); translator.close() }
          }
          .addOnFailureListener { error -> promise.reject("MODEL_DOWNLOAD_FAILED", error.message, error); translator.close() }
      }
    }
  }
}

File: modules/on-device-translation/ios/OnDeviceTranslationModule.swift
import ExpoModulesCore
import MLKitTranslate
import MLKitCommon

public class OnDeviceTranslationModule: Module {
  public func definition() -> ModuleDefinition {
    Name("OnDeviceTranslation")

    AsyncFunction("translate") { (text: String, source: String, target: String, promise: Promise) in
      guard let from = TranslateLanguage.fromLanguageTag(source),
            let to = TranslateLanguage.fromLanguageTag(target) else {
        promise.reject("UNSUPPORTED_LANGUAGE", "Deze taal wordt niet ondersteund op dit toestel")
        return
      }
      let translator = Translator.translator(options: TranslatorOptions(sourceLanguage: from, targetLanguage: to))
      let conditions = ModelDownloadConditions(allowsCellularAccess: false, allowsBackgroundDownloading: false)
      translator.downloadModelIfNeeded(with: conditions) { error in
        if let error = error {
          promise.reject("MODEL_DOWNLOAD_FAILED", error.localizedDescription)
          return
        }
        translator.translate(text) { result, error in
          if let result = result { promise.resolve(result) }
          else { promise.reject("TRANSLATION_FAILED", error?.localizedDescription ?? "Vertalen mislukt") }
        }
      }
    }
  }
}
